java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.qhelp

```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+If a <code>HostnameVerifier</code> always returns <code>true</code> it will not verify the hostname at all.
+This stops Transport Layer Security (TLS) providing any security and allows an attacker to perform a Man-in-the-middle attack against the application.
+</p>
+
+<p>
+An attack might look like this:
+</p>
+
+<ol>
+  <li>The program connects to <code>https://example.com</code>.</li>
+  <li>The attacker intercepts this connection and presents one of their valid certificates they control, for example one from Let's Encrypt.</li>
+  <li>Java verifies that the certificate has been issued by a trusted certificate authority.</li>
```

Comment:
We don't normally talk about a programming language as performing tasks. I'm not entirely sure what the right change would be in these steps. Possibly "The program", as in step 1?

Reply:
Thank you for explaining. It's much clearer to me now. I'd value someone else's thoughts on whether just using "Java" is fine in this instance, or whether we should change it to something like: "The Java environment of the program verifies..."

Reply:
@aschackmull do you maybe have some input for this?

Reply:
Well the Java language as such of course doesn't know anything about connections or certificates or the like, so I agree that the sentence sounds weird. Presumably it is the library in which we plug in a custom HostnameVerifier that is doing the verification? I guess some library code in the package `javax.net.ssl`? We should probably refer to that library as the subject in this sentence (assuming I've understood things correctly - I'm not familiar with these libraries).

Reply:
```suggestion
  <li>The `TrustManager` of the program verifies that the certificate has been issued by a trusted certificate authority.</li>
```